File: lib/api/cyclone.ts
import { apiRequest } from './client';

export interface CycloneDataResponse {
  status: 'CONNECTED' | 'LOADING' | 'DEGRADED' | 'ERROR' | 'UNAVAILABLE';
  source: string;
  dataset: string;
  active_cyclones_count: number;
  retrieved_at: string;
  data: {
    status_message?: string;
    basin_state?: string;
    tracks?: any[];
    wind_radii?: any;
    cone_of_uncertainty?: any;
  } | null;
  error?: string | null;
}

export async function fetchActiveCyclones(): Promise<CycloneDataResponse> {
  try {
    return await apiRequest<CycloneDataResponse>('cyclones/active');
  } catch (err: any) {
    return {
      status: 'UNAVAILABLE',
      source: 'IMD (Cyclone Warning Division)',
      dataset: 'North Indian Ocean Tropical Cyclone Advisory',
      active_cyclones_count: 0,
      retrieved_at: new Date().toISOString(),
      data: {
        status_message: 'Unable to connect to IMD Cyclone Warning Division endpoint.',
        basin_state: 'UNKNOWN',
        tracks: []
      },
      error: err.message
    };
  }
}

File: backend/app/providers/imd/cyclone.py
"""
ORCA Backend — IMD Cyclone Data Provider
Integrates official IMD Cyclone Warning Division endpoints:
- /cyclone_track (Observed & Forecast Position History)
- /cyclone_wind (Gale/Squall Wind Radii Polygons)
- /cyclone_cou (Cone of Uncertainty Envelope)
"""
import logging
from typing import Optional, Dict, Any, List
from datetime import datetime, timezone

from app.providers.imd.client import imd_client

logger = logging.getLogger("orca.imd.cyclone")


async def get_active_cyclones() -> Dict[str, Any]:
    """
    Fetch active cyclone tracks and bulletins from IMD Cyclone Warning Division.
    Returns GeoJSON-ready track, cone of uncertainty, and gale wind radii.
    """
    now_iso = datetime.now(tz=timezone.utc).isoformat()
    
    # 1. Query official IMD cyclone track endpoint
    res_track = await imd_client.get("cyclone_track", ttl_seconds=600)
    res_wind = await imd_client.get("cyclone_wind", ttl_seconds=600)
    res_cou = await imd_client.get("cyclone_cou", ttl_seconds=600)

    if res_track["status"] == "CONNECTED" and res_track["data"]:
        return {
            "status": "CONNECTED",
            "source": "IMD (Cyclone Warning Division, New Delhi)",
            "dataset": "North Indian Ocean Tropical Cyclone Advisory",
            "active_cyclones_count": len(res_track["data"]) if isinstance(res_track["data"], list) else 1,
            "retrieved_at": res_track["retrieved_at"],
            "data": {
                "tracks": res_track["data"],
                "wind_radii": res_wind.get("data"),
                "cone_of_uncertainty": res_cou.get("data")
            },
            "error": None
        }

    # If no active tropical cyclone bulletin is issued by IMD
    return {
        "status": "CONNECTED",
        "source": "IMD (Cyclone Warning Division)",
        "dataset": "North Indian Ocean Tropical Cyclone Advisory",
        "active_cyclones_count": 0,
        "retrieved_at": now_iso,
        "data": {
            "status_message": "No active tropical cyclone / deep depression in the North Indian Ocean basin (Arabian Sea & Bay of Bengal).",
            "basin_state": "QUIET / NORMAL VORTICITY",
            "tracks": [],
            "wind_radii": None,
            "cone_of_uncertainty": None
        },
        "error": None
    }
